iceoryx_hoofs/test/moduletests/test_cxx_functional_interface_and_then.cpp

```diff
@@ -0,0 +1,143 @@
+// Copyright (c) 2022 by Apex.AI Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+#include "test_cxx_functional_interface_types.hpp"
+
+namespace
+{
+using namespace test_cxx_functional_interface;
+using namespace ::testing;
+
+template <bool HasValue>
+struct AndThenIsCalledCorrectlyWhenValid;
+
+template <>
+struct AndThenIsCalledCorrectlyWhenValid<false>
+{
+    template <typename TestFactory, typename AndThenCall>
+    static void performTest(const AndThenCall& andThenCall)
+    {
+        auto sut = TestFactory::createValidObject();
+        bool wasCallbackCalled = false;
+        andThenCall(sut, [&] { wasCallbackCalled = true; });
+        EXPECT_TRUE(wasCallbackCalled);
+    }
+};
+
+template <>
+struct AndThenIsCalledCorrectlyWhenValid<true>
+{
+    template <typename TestFactory, typename AndThenCall>
+    static void performTest(const AndThenCall& andThenCall)
+    {
+        auto sut = TestFactory::createValidObject();
+        bool wasCallbackCalled = false;
+        andThenCall(sut, [&](auto& arg) {
+            wasCallbackCalled = true;
+            EXPECT_EQ(arg, TestFactory::usedTestValue);
+        });
+        EXPECT_TRUE(wasCallbackCalled);
+    }
+};
+
+TYPED_TEST(FunctionalInterface_test, AndThenIsCalledCorrectlyWhenValid_LValueCase)
+{
+    using SutType = typename TestFixture::TestFactoryType::Type;
+    AndThenIsCalledCorrectlyWhenValid<iox::cxx::internal::HasValueMethod<SutType>::value>::template performTest<
+        typename TestFixture::TestFactoryType>([](auto& sut, auto callback) { sut.and_then(callback); });
```

Comment:
Just want to make sure that I get it right:
1. `HasValueMethod` defined in `functional_interface.hpp` evaluates to true/false depending on whether `SutType` has a `value()` method or not. The specialization of `AndThenIsCalledCorrectlyWhenValid` is selected accordingly.
2. `performTest` is called with `[](auto& sut, auto callback) { sut.and_then(callback); }` as argument
3. in `performTest` a valid object is created (=`sut`)
4. the callback argument `andThenCall` is executed like that:
```
sut.and_then([&](auto& arg) {
    wasCallbackCalled = true;
    EXPECT_EQ(arg, TestFactory::usedTestValue);
});
```
with `sut` from 3. and `AndThenIsCalledCorrectlyWhenValid<true>`.

Please let me know if my understanding is correct.